### PR TITLE
finalizer for OBC cleanup

### LINKED ISSUE #4486: Add a finalizer to the operator r/t the removal of RGW OBCs
**Is this a bug report or feature request?**
* Feature Request

**What should the feature do:**

At present, Rook has a finalizer that checks for the removal of PVCs, but no such finalizer exists for OBCs created through RGW. In order to fully remove Rook and all managed resources, such a finalizer should be added.

**What is use case behind this feature:**

As a Rook user, I want Rook to ensure that all of its managed resources (including OBCs) are removed when I uninstall it, so that my k8s cluster does not contain unmanaged orphan resources.

**Environment**:

Any/all.

### PR DESCRIPTION
**Pull request by sp98**
**Description of your changes:**
- This PR is for rebasing #4683 with master. With controller run time changes, most of the changes in #4683 got redundant. Only change is waiting for all the object buckets to be deleted before removing the finalizer. 
- When deleting object store, reconciler will check if all the object buckets are cleaned up or not.
- If not, the reconcile will fail and keep on trying again. 
- Once the buckets are deleted, reconcile operation will succeed and continue to delete the object store.  

**Which issue is resolved by this Pull Request:**
Resolves #4486

**Checklist:**

- [ ] **CommitLint Bot**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/master/development-flow.html#commit-structure).
- [ ] **Skip Tests for Docs**: Add the flag for skipping the build if this is only a documentation change. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for the flag.
- [ ] **Skip Unrelated Tests**: Add a flag to run tests for a specific storage provider. See [test options](https://github.com/rook/rook/blob/master/INSTALL.md#test-storage-provider).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/master/development-flow.html#submitting-a-pull-request)
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.

[test ceph]

### CODE REVIEW

**Comment by sp98** [comment]
Looking into the CI failures.

File: pkg/operator/ceph/object/controller.go [hunk 1]
```diff
@@ -319,3 +338,36 @@ func updateStatus(client client.Client, name types.NamespacedName, status string
...
+			logger.Infof("no objectbuckets for objectstore %q to clean up.", objectstore.Namespace)
+			break
+		}
+
+		retryCount++
+		if retryCount == objectstoreDeleteMaxRetries {
+			bucketNames := make([]string, 0)
+			for _, bucket := range objectBuckets.Items {
+				bucketNames = append(bucketNames, bucket.Name)
+			}
```
Comment (thotz) [inline pkg/operator/ceph/object/controller.go:360]: @leseb : May be it is very naive question, if I understand correctly here object store is getting deleted. And we trying fetch info from obcs to see buckets is getting deleted. IMHO obc/ob will still present even if objectstore is deleted.
```console
# kubectl delete -f cluster/examples/kubernetes/ceph/object-test-sam.yaml 
cephobjectstore.ceph.rook.io "my-store" deleted

# kubectl get obc
NAME                  AGE
ceph-delete-bucket    3m46s
ceph-delete-bucket2   85s

# kubectl get ob
NAME                              AGE
obc-default-ceph-delete-bucket    3m52s
obc-default-ceph-delete-bucket2   92s
```

Comment (leseb) [reply pkg/operator/ceph/object/controller.go:360]: Yes, it will but its existence does not make much sense anymore. The buckets are not accessible anymore and the data are gone.

It's just not clear from that code where the deletion of OBC is happening though...

Comment (sp98) [reply pkg/operator/ceph/object/controller.go:360]: There is a check in the yaml file - "preservePoolsOnDelete". When this is false, rook deletes the Object Pools [here](https://github.com/rook/rook/blob/1ea73ce1233fcaa4a1b5518967eef2dcbd0cade6/pkg/operator/ceph/object/objectstore.go#L77). And deleting the pools would delete the object buckets, is what I read [here](https://github.com/rook/rook/pull/4683#issuecomment-575315487). 
So we will only wait for the object buckets to be deleted when:
1. "preservePoolsOnDelete" is set to false. ( That's in this PR)
2.  when the cleanup policy is set for the deletion of Object buckets. (Planned for another PR)

File: pkg/operator/ceph/object/controller.go [hunk 2]
```diff
@@ -203,12 +213,24 @@ func (r *ReconcileCephObjectStore) reconcile(request reconcile.Request) (reconci
 	// DELETE: the CR was deleted
 	if !cephObjectStore.GetDeletionTimestamp().IsZero() {
 		logger.Debugf("deleting store %q", cephObjectStore.Name)
+		cephObjectStore.Spec.PreservePoolsOnDelete = false
```
Comment (travisn) [inline pkg/operator/ceph/object/controller.go:216]: Was this only for testing? I assume we're not committing this change to always purge the pools.

File: pkg/operator/ceph/object/controller.go [hunk 3]
```diff
@@ -203,12 +213,24 @@ func (r *ReconcileCephObjectStore) reconcile(request reconcile.Request) (reconci
...
 		cfg := clusterConfig{context: r.context, store: cephObjectStore}
 		err := cfg.deleteStore()
 		if err != nil {
 			return reconcile.Result{}, errors.Wrapf(err, "failed to delete store %q. ", cephObjectStore.Name)
 		}
 
+		// TODO: Add cleanupPolicy for confirming deletion of object buckets.
+		// TODO: Always set PreservePoolsOnDelete to false if object buckets cleanupPolicy is present.
+		if !cephObjectStore.Spec.PreservePoolsOnDelete {
```
Comment (travisn) [inline pkg/operator/ceph/object/controller.go:225]: The PreservePoolsOnDelete setting seems sufficient for whether to delete the pools. I would say we don't need either of these TODOs.
- The cleanupPolicy is for the cluster-wide settings for mons and osds when you're destroying the entire cluster. Here we aren't destroying an entire cluster, we are just deleting an object store.
- If the cleanupPolicy in the cluster CR is set to destroy the cluster, it doesn't seem like we need to worry about that. The pool finalizer can still just worry about its own settings.

File: pkg/operator/ceph/object/controller.go [hunk 4]
```diff
@@ -203,12 +213,24 @@ func (r *ReconcileCephObjectStore) reconcile(request reconcile.Request) (reconci
...
+			// Wait for all the object buckets in an object store to be deleted
+			err = r.waitForObjectBucketCleanup(cephObjectStore)
```
Comment (travisn) [inline pkg/operator/ceph/object/controller.go:227]: Although if the cleanupPolicy is set, we really shouldn't wait for object bucket cleanup since the intent is to destroy everything.

File: pkg/operator/ceph/object/controller.go (hunk 1 above)
Comment (thotz) [reply pkg/operator/ceph/object/controller.go:362]: Let me rephrase my concern, I agree that if you delete the pool, all the (actual)buckets will be removed but obc's still may exists.  And here for checking, the bucket name is fetched from OBC, which was confusing for me.

Comment (sp98) [reply pkg/operator/ceph/object/controller.go:362]: I was wrong in my comment [above](https://github.com/rook/rook/pull/5436#discussion_r423681260).  Object buckets are still there after the pools are deleted (like you mentioned [here](https://github.com/rook/rook/pull/5436#discussion_r423658160)). So current changes won't work. I'm working on fixing that now.

File: pkg/operator/ceph/object/controller.go [hunk 5]
```diff
@@ -319,3 +340,81 @@ func updateStatus(client client.Client, name types.NamespacedName, status string
...
+	objectBuckets, err := r.bktclient.ObjectbucketV1alpha1().ObjectBuckets().List(metav1.ListOptions{LabelSelector: selector})
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get buckets for objectstore %q in namespace %q", objectstore.Name, objectstore.Namespace)
+	}
+	return objectBuckets, nil
+}
+
+func (r *ReconcileCephObjectStore) GetObjectBucketClaims(objectstore *cephv1.CephObjectStore) (*bktv1alpha1.ObjectBucketClaimList, error) {
+	selector := fmt.Sprintf("bucket-provisioner=%s", bucketProvisioner)
+	objectBucketsClaims, err := r.bktclient.ObjectbucketV1alpha1().ObjectBucketClaims("default").List(metav1.ListOptions{LabelSelector: selector})
```
Comment (sp98) [inline pkg/operator/ceph/object/controller.go:384]: @travisn @leseb I can't seem to find a way on how not to hard code the `namespace` to get the existing `ObjectBucketClaims`.  Any inputs?

File: pkg/operator/ceph/object/controller.go (hunk 3 above)
Comment (sp98) [reply pkg/operator/ceph/object/controller.go:218]: If I understood the requirement correctly, we need a clean uninstall. When the cluster is deleted (with or without the cleanupPolicy) the OBCs and OBs are still around and we wanted to delete these orphaned resources as well.

File: pkg/operator/ceph/object/controller.go (hunk 5 above)
Comment (leseb) [reply pkg/operator/ceph/object/controller.go:384]: I recently introduced a change to watch on the cluster namespace by default. So perhaps you can rely on that since all newly created OBC will work this way.
For backward compatibility, I think hardcoding default is fine though.

File: pkg/operator/ceph/object/controller.go [hunk 6]
```diff
@@ -47,7 +48,8 @@ import (
 )
 
 const (
-	controllerName = "ceph-object-controller"
+	controllerName    = "ceph-object-controller"
+	bucketProvisioner = "rook-ceph.ceph.rook.io-bucket"
```
Comment (leseb) [inline pkg/operator/ceph/object/controller.go:52]: We cannot hardcode this the `rook-ceph` part is the namespace. This won't always be the case.
Just keep `.ceph.rook.io-bucket `

File: pkg/operator/ceph/object/controller.go [hunk 7]
```diff
@@ -319,3 +329,32 @@ func updateStatus(client client.Client, name types.NamespacedName, status string
...
+func (r *ReconcileCephObjectStore) handleDelete(objectstore *cephv1.CephObjectStore) error {
+	//  Wait for all the object buckets to be cleaned up
+	err := r.waitForObjectBucketCleanup(objectstore)
+	if err != nil {
+		return errors.Wrap(err, "failed to cleanup buckets")
+	}
+	return nil
+}
+
+func (r *ReconcileCephObjectStore) waitForObjectBucketCleanup(objectstore *cephv1.CephObjectStore) error {
```
Comment (leseb) [inline pkg/operator/ceph/object/controller.go:342]: I think the name of the method is not accurate anymore. We don't wait.

File: pkg/operator/ceph/object/controller.go [hunk 8]
```diff
@@ -319,3 +329,32 @@ func updateStatus(client client.Client, name types.NamespacedName, status string
 	}
 	logger.Debugf("object store %q status updated to %q", name, status)
 }
+
+func (r *ReconcileCephObjectStore) handleDelete(objectstore *cephv1.CephObjectStore) error {
+	//  Wait for all the object buckets to be cleaned up
+	err := r.waitForObjectBucketCleanup(objectstore)
+	if err != nil {
+		return errors.Wrap(err, "failed to cleanup buckets")
```
Comment (leseb) [inline pkg/operator/ceph/object/controller.go:337]: I don't think this is the error here, we fail because there are OB still present

File: pkg/operator/ceph/object/controller.go (hunk 6 above)
Comment (leseb) [reply pkg/operator/ceph/object/controller.go:52]: Also, you need to add this https://github.com/rook/rook/blob/master/pkg/operator/ceph/object/bucket/util.go#L47 for backward compatibility, please export it and re-use it here.

File: pkg/operator/ceph/object/controller.go [hunk 9]
```diff
@@ -319,3 +329,32 @@ func updateStatus(client client.Client, name types.NamespacedName, status string
...
+	}
+	return nil
+}
+
+func (r *ReconcileCephObjectStore) waitForObjectBucketCleanup(objectstore *cephv1.CephObjectStore) error {
+	selector := fmt.Sprintf("bucket-provisioner=%s", bucketProvisioner)
```
Comment (leseb) [inline pkg/operator/ceph/object/controller.go:343]: You need the namespace to build it, just like in https://github.com/rook/rook/blob/master/pkg/operator/ceph/object/bucket/util.go#L60-L67

File: pkg/operator/ceph/object/controller.go [hunk 10]
```diff
@@ -319,3 +329,32 @@ func updateStatus(client client.Client, name types.NamespacedName, status string
...
+	objectBuckets, err := r.bktclient.ObjectbucketV1alpha1().ObjectBuckets().List(metav1.ListOptions{LabelSelector: selector})
+	if err != nil {
+		return errors.Wrapf(err, "failed to list buckets for objectstore %q in namespace %q", objectstore.Name, objectstore.Namespace)
+	}
+
+	if len(objectBuckets.Items) == 0 {
+		logger.Infof("no buckets for objectstore %q in namespace %q to clean up", objectstore.Name, objectstore.Namespace)
```
Comment (leseb) [inline pkg/operator/ceph/object/controller.go:350]: we never clean up

File: pkg/operator/ceph/object/controller.go [hunk 11]
```diff
@@ -319,3 +329,34 @@ func updateStatus(client client.Client, name types.NamespacedName, status string
 	}
 	logger.Debugf("object store %q status updated to %q", name, status)
 }
+
+func (r *ReconcileCephObjectStore) handleDelete(objectstore *cephv1.CephObjectStore) error {
+	// verify if all the object buckets are cleaned up
+	err := r.verifyObjectBucketCleanup(objectstore)
+	if err != nil {
+		return err
```
Comment (leseb) [inline pkg/operator/ceph/object/controller.go:337]: Please name the error.

File: pkg/operator/ceph/object/objectstore.go [hunk 12]
```diff
@@ -333,3 +337,17 @@ func poolName(storeName, poolName string) string {
 	// the name of the pool is <instance>.<name>, except for the pool ".rgw.root" that spans object stores
 	return fmt.Sprintf("%s.%s", storeName, poolName)
 }
+
+// GetObjectBucketProvisioner returns the bucket provisioner name appended with operator namespace if OBC is watching on it
+func GetObjectBucketProvisioner(c *clusterd.Context, namespace string) string {
```
Comment (leseb) [inline pkg/operator/ceph/object/objectstore.go:342]: Please add unit tests

Comment (sp98) [reply pkg/operator/ceph/object/objectstore.go:342]: added unit tests.

File: pkg/operator/ceph/object/controller.go [hunk 13]
```diff
@@ -319,3 +329,34 @@ func updateStatus(client client.Client, name types.NamespacedName, status string
 	}
 	logger.Debugf("object store %q status updated to %q", name, status)
 }
+
+func (r *ReconcileCephObjectStore) handleDelete(objectstore *cephv1.CephObjectStore) error {
```
Comment (leseb) [inline pkg/operator/ceph/object/controller.go:333]: Is this method serving a real purpose? What about just calling `verifyObjectBucketCleanup` in L211?

Comment (sp98) [reply pkg/operator/ceph/object/controller.go]: removed `handleDelete` function.

File: pkg/operator/ceph/object/controller.go [hunk 14]
```diff
@@ -203,10 +207,16 @@ func (r *ReconcileCephObjectStore) reconcile(request reconcile.Request) (reconci
 	// DELETE: the CR was deleted
 	if !cephObjectStore.GetDeletionTimestamp().IsZero() {
 		logger.Debugf("deleting store %q", cephObjectStore.Name)
+
+		err := r.verifyObjectBucketCleanup(cephObjectStore)
+		if err != nil {
+			return reconcile.Result{}, errors.Wrapf(err, "failed to delete store %q", cephObjectStore.Name)
```
Comment (travisn) [inline pkg/operator/ceph/object/controller.go:213]: The operator should continue attempting to delete the object store by re-queueing the event. So instead of returning an empty result, we need a re-queue event here.

Comment (sp98) [reply pkg/operator/ceph/object/controller.go:213]: Returning error in the `Reconcile` would re-queue the event and would try to delete the object again.

Comment (travisn) [reply pkg/operator/ceph/object/controller.go:213]: If the err is returned, you'll get immediate retries with exponential backoff. The reason I'd suggest a requeue event is so we can control the retry interval. In this case we know it can take time for the buckets to be deleted by the admin, so the exponential backoff doesn't make as much sense.

Comment (sp98) [reply pkg/operator/ceph/object/controller.go]: Thanks for pointing that out and sharing a reference on how to do that. I've updated the code to add retry interval for reconcile.
